README.md
# Пояснения к решению
## Решение состоит из трех единиц компиляции, каждая из них представлена в двух форматах:
- Код C
- Комментированный, оптимизированный код Assembly
# Их спискок:
1. Main:
>  * Main.c
>  * Main.s
 
2. Get-rand-num-func:
>  * Get-rand-num-func.c
>  * Get-rand-num-func.s
3. func:
>  * func.c
>  * func.s
<br />

## В решении присутствуют внешние текстовые файлы:

1. times.txt - Выводится длительность каждого из тестов при запуске режима тестирования
2. input.txt - Обычный файл ввода
3. tests.txt - Cписок тестов
4. output.txt - Обычный файловый вывод
<br />


# Для проверки решения, вся документация вынесена в отдельную папку info
1. [Список изменений](https://github.com/zemlianin/first-HW-ABC/blob/master/Info/Changes.md)
2. [Отчет](https://github.com/zemlianin/first-HW-ABC/blob/master/Info/Report.md)

# Инструкция
Exe-файл создается за счет команды линковки трех файлов
```
gcc ./Main.o ./func.o ./get-rand-num-func.c -o ./solution.exe

```
1. Во время запуска программы пользователю требуется по средствам параметров запуска ввести параметр рандома.
```
./solution.exe 12
```
2. Следующим шагом пользователю требуется ввести размер массива А (При использовании тестов записанных в файл, требует указывать размер 10, тк там лежат массивы размера 10)
3. Дальнейшее взаимодействие с программой происходит за счет вывода соответствующих сообщений в консоли


Info/Changes.md
# Список изменений
### 1. Удаление ненужных перезаписываний регистров, таких как:

``` 
lea	rax, .LC2[rip]
mov	rsi, rax
  ```
  ```
lea	rax, .LC3[rip] 
mov	rdi, rax
  ```
  ```
lea	rax, .LC1[rip]
mov	rdi, rax
```
```
lea	rax, .LC0[rip]
mov	rdi, rax
  ```
  
  ```
  mov	rax, QWORD PTR -24[rbp]
lea	rcx, .LC1[rip]
mov	rsi, rcx
mov	rdi, rax
  ```
  
  ```
  add	rax, rdx
mov	rsi, rax
  ```
  
  ```
  lea	rax, .LC1[rip]
mov	rdi, rax
  ```
  
  ```
  lea	rax, .LC6[rip]
mov	rdi, rax
  ```
  
  ```
  lea	rax, -860[rbp]
mov	rsi, rax
  ```
  
  ```
  lea	rax, -865[rbp]
mov	rdi, rax
  ```
  
  ```
  lea	rax, .LC4[rip]
mov	rsi, rax
lea	rax, .LC5[rip]
mov	rdi, rax
  ```
  
  ```
  mov	rax, QWORD PTR -24[rbp]
lea	rcx, .LC1[rip]
mov	rsi, rcx
mov	rdi, rax
  ```
  
  ```
mov	rax, QWORD PTR -32[rbp]
lea	rcx, .LC1[rip]
mov	rsi, rcx
mov	rdi, rax
```

```
mov	rax, QWORD PTR -32[rbp]
mov	rsi, rax
```

```
lea	rcx, -852[rbp]
lea	rax, -448[rbp]
mov	rsi, rcx
mov	rdi, rax
  ```
  
 ### Вместо этого, строка/адрес/переменная записываются напрямую в конечных регистр, без посредника rax или rcx
 ### 2. Удаление кода сгенерированного GCC
 ```
 	.ident	"GCC: (Ubuntu 11.2.0-19ubuntu1) 11.2.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	1f - 0f
	.long	4f - 1f
	.long	5
0:
	.string	"GNU"
1:
	.align 8
	.long	0xc0000002
	.long	3f - 2f
2:
	.long	0x3
3:
	.align 8
4:
```
### 3. Повсеместное удаление макроса cdqe
### 4. Удаление макроса endbr64
### 5. Оптимизация за счет хранения некторых переменных стека в регистре, среди которых:
## Main.s:
* arcs -> r8d
* argv[] -> r9
* *input -> r12
* begin -> r13
* end -> r14
* i - неудается заменить, Segmentation fault (core dumped)
* command - неудается заменить, Segmentation fault (core dumped)
* size - неудается заменить, Segmentation fault (core dumped)
## func.s:
* *A -> r14
* *B -> r15
* size -> неудается заменить, Segmentation fault (core dumped)



Main-full.s
# Delete information from gcc
# Delete name 
# Удаление endbr64
# Удаление посредника раикс при получении значения переменной

	.intel_syntax noprefix
	.text
	.section	.rodata 		#  Начало секции, объявление строк
.LC0: 
	.string	"input size"
.LC1:
	.string	"%d"
	.align 8
.LC2:
	.string	"input command: \n 1 - read from file \n 2 - read from console\n 3 - Random\n 0 - Testing"
.LC3:
	.string	"%s"
.LC4:
	.string	"r"
.LC5:
	.string	"tests.txt"
.LC6:
	.string	"input range of numbers"
.LC7:
	.string	"w"
.LC8:
	.string	"times.txt"
.LC9:
	.string	"Sum of Time in ms: "
.LC10:
	.string	"output.txt"
	.align 8
.LC11:
	.string	"output command: \n 1 - write B to file \n 2 - write B to console"
	.text
	.globl	main 				# Объявление символа мэйн, обозначение его как функции
	.type	main, @function
main:
	push	rbp 				# Запись рбп в стэк и запись рсп в рбп, начало нового метода 
	mov	rbp, rsp
	sub	rsp, 896
	
	mov	DWORD PTR -884[rbp], edi 	# Получение параметра размера массива
	mov	QWORD PTR -896[rbp], rsi 	# Получение параметра указателя на массив

	lea	rdi, .LC0[rip] 			# Запись строки в регистр
	call	puts@PLT 			# Запуск метода принт, вывод строки из регистра
	
	lea	rsi, -852[rbp] 			# Запись сайз в регистр
	lea	rdi, .LC1[rip] 			# Запись строки в регистр
	mov	eax, 0 				# Обнуление регистра
	call	__isoc99_scanf@PLT 		# Запуск метода скан
	
	lea	rsi, .LC2[rip]
	lea	rdi, .LC3[rip]
	mov	eax, 0
	call	printf@PLT 			# Запись строк, параметров принта в регистры и вызов принт
	
	lea	rsi, -856[rbp]
	lea	rdi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT  		# Запись строк, параметров принта в регистры и вызов метода скан
	
	mov	eax, DWORD PTR -856[rbp] 	# Запись переменной команда в регистр
	cmp	eax, 1 				# Сравнение с единицей
	jne	.L2				# Если Они не равны прыжок в точку 2
	
	lea	rsi, .LC4[rip]
	lea	rdi, .LC5[rip]
	call	fopen@PLT 			# Запись строк как параметров в регистры и вызов метода открытия потока\
	
	mov	QWORD PTR -24[rbp], rax 
	mov	DWORD PTR -4[rbp], 0 # i = 0
	jmp	.L3
.L4:
	lea	rdx, -448[rbp] 			# Запись адреса А в регистр
	mov	eax, DWORD PTR -4[rbp] 		#  Запись параметра в регистр
	sal	rax, 2
	add	rdx, rax   			#  Расчет адреса текущего элемента массива
	
	mov	rdi, QWORD PTR -24[rbp] 	# Запись параметров метода скан в регистры, строка и указатель
	lea	rsi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_fscanf@PLT 		#  Запуск метода скан
	
	add	DWORD PTR -4[rbp], 1 		# Прибавление 1 к параметру
.L3:
	mov	eax, DWORD PTR -852[rbp]	# Проверка условия выхода из цикла
	cmp	DWORD PTR -4[rbp], eax
	jl	.L4				# Если параметр меньше сайз, прыжок в точку 4
	
	
.L2:
	mov	eax, DWORD PTR -856[rbp] 	# Запись перемнной команда в регистр
	cmp	eax, 2 				# Сравнение команды с двойкой
	jne	.L5 				#  Если условие не выполнено прыжок в точку 5
	mov	DWORD PTR -4[rbp], 0		# Запись параметра цикла в регистр
	jmp	.L6				#  Прыжок в точку 6
.L7:
	lea	rdx, -448[rbp] 			# Получение адреса начала массива А
	mov	eax, DWORD PTR -4[rbp] 		# Запись параметра цикла в регистр
	
	sal	rax, 2 				# расчет адреса текущего элемента массива
	add	rsi, rdx
	
	lea	rdi, .LC1[rip] 			# Запись строки в регистр
	mov	eax, 0
	call	__isoc99_scanf@PLT 		# Запуск метода скан
	
	add	DWORD PTR -4[rbp], 1  		# Увеличение параметра на 1
.L6:
	mov	eax, DWORD PTR -852[rbp] 	#  Запись значения сайз в регистр
	cmp	DWORD PTR -4[rbp], eax 		#  Сравнение значения сайз с параметром
	jl	.L7				#  Если параметр меньше, прыжок в точку 7
.L5:
	mov	eax, DWORD PTR -856[rbp] 	# Запись команды в регистр
	cmp	eax, 3				# Сравнение команды с 3
	jne	.L8				# Если условие не выполнено то прыжок вточку 8
	
	lea	rdi, .LC6[rip]			# Запись Строки в регистр и вызов метода принт для строк
	call	puts@PLT	
			
	lea	rsi, -860[rbp]			# Запись адреса переменной диапазон значений в регистр
	lea	rdi, .LC1[rip]			# Запись строки в регистр
	mov	eax, 0
	call	__isoc99_scanf@PLT		# Вызов метода скан
	
	cmp	DWORD PTR -884[rbp], 1		# Сравнение парамметра мметода мэйн с 1
	jne	.L9				
	mov	BYTE PTR -865[rbp], 48		# Запись нолика в переменную параметр  рандома
	jmp	.L10
.L9:
	mov	rax, QWORD PTR -896[rbp]	# Получение ссылки на массив
	mov	rax, QWORD PTR [rax]		# Получение элемента массива
	movzx	eax, BYTE PTR 1[rax]
	mov	BYTE PTR -865[rbp], al		# Запись полученного значения в перемменную
.L10:
	mov	edx, DWORD PTR -860[rbp]	# Запись параметров метода ранд в регистры
	mov	esi, DWORD PTR -852[rbp]
	lea	rcx, -448[rbp]
	lea	rdi, -865[rbp]
	call	Rand@PLT			# Запуск метода ранд
.L8:
	mov	eax, DWORD PTR -856[rbp]	# Запись переменной команд в регистр
	test	eax, eax			# Сравнение переменной с нулем через побитовое и
	jne	.L11				# Если условие не выполнено прыжок в точку 11
	
	mov	eax, DWORD PTR -852[rbp]	# Запись сайз в сайзН
	mov	DWORD PTR -864[rbp], eax
		
	lea	rsi, .LC4[rip]			# Задание параметров и открытие потока на чтение
	lea	rdi, .LC5[rip]
	call	fopen@PLT
	
	mov	QWORD PTR -24[rbp], rax	
		
	lea	rsi, .LC7[rip]			# Открытие потока на запись
	lea	rdi, .LC8[rip]
	call	fopen@PLT
	
	mov	QWORD PTR -32[rbp], rax
	mov	DWORD PTR -4[rbp], 0		# Присвоение параметру нуля
	jmp	.L12				# Прыжок в точку 12
.L15:
	mov	DWORD PTR -8[rbp], 0		# Присвоение параметру внутреннего цикла нуля
	jmp	.L13				# Прыжок в точку 13
.L14:
	lea	rdx, -448[rbp]			# Получение адреса Начала массива А
	
	mov	eax, DWORD PTR -4[rbp]		# Расчет Адреса текущего элемента массива
	sal	rax, 2
	add	rdx, rax
	
	mov	rdi, QWORD PTR -24[rbp]		# Запуск метода скан
	lea	rsi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_fscanf@PLT
	
	add	DWORD PTR -8[rbp], 1		# +1 к внутреннему параметру
.L13:
	mov	eax, DWORD PTR -852[rbp]	# Запись значения сайз в регистр
	cmp	DWORD PTR -8[rbp], eax		# Сравнение внутреннего параметра с сайзом
	jl	.L14				# Если параметр меньше прыжок в точку 14
	
	call	clock@PLT			# Вызов метода клок
	mov	QWORD PTR -40[rbp], rax		# Запись результата выполнения функции в стэк
	
	lea	rdx, -848[rbp]			# Вызов функции таск
	lea	rsi, -864[rbp]			# Запись переменных в регистры
	lea	rdi, -448[rbp]
	call	Task@PLT
	
	call	clock@PLT			# Вызов функции клок
	mov	QWORD PTR -48[rbp], rax		# Запись результата в стэк
	
	mov	rax, QWORD PTR -48[rbp]		# Запись конечного момента времени в регистр
	mov	edx, eax			
	mov	rax, QWORD PTR -40[rbp]		# Запись начального момента времени в регистр
	mov	ecx, eax
	mov	eax, edx
	sub	eax, ecx			# Получение разности значений
	mov	edx, eax			# Запись результата в регистр
	
	mov	rdi, QWORD PTR -32[rbp]		# Вызов функции эф принт 
	lea	rsi, .LC1[rip]
	mov	eax, 0
	call	fprintf@PLT
	
	mov	rax, QWORD PTR -48[rbp]		# Запись конечного момента времени в регистр
	mov	edx, eax
	mov	rax, QWORD PTR -40[rbp]		# Запись начального момента времени в регистр
	mov	ecx, eax
	mov	eax, edx
	sub	eax, ecx			# Получение Разности моментов времени
	
	add	DWORD PTR -12[rbp], eax		# Прибавление нового отрезка времени к существующей сумме
	
	mov	rsi, QWORD PTR -32[rbp]		# Запись указателя на поток в регистр
	mov	edi, 10				# Запись символа переноса строки в регистр
	call	fputc@PLT			# Вызов метода записи в файл
	
	add	DWORD PTR -4[rbp], 1		# Внешний параметр +=1
.L12:
	cmp	DWORD PTR -4[rbp], 9 		# Сравнивание параметра с 9
	jle	.L15				# Если меньше то прыжок в точку 15
	
	lea	rsi, .LC9[rip]			# Запись параметров в регистры, вызов метода принт
	lea	rdi, .LC3[rip]
	mov	eax, 0
	call	printf@PLT
	
	mov	esi, DWORD PTR -12[rbp]		# Запись суммы времени на тесты в регистр
	lea	rdi, .LC1[rip]			# Запись строки в регситр
	mov	eax, 0
	call	printf@PLT			# Вызов принт
	mov	eax, 0
	
	jmp	.L23				# Прыжок к концу программы
.L11:

	lea	rdx, -848[rbp]			# Запись параметров в регисты, вызов метода таск
	lea	rsi, -852[rbp]
	lea	rdi, -448[rbp]
	call	Task@PLT
	
	lea	rsi, .LC7[rip]			# Запись параметров в регистры 
	lea	rdi, .LC10[rip]
	call	fopen@PLT			# Вызов метода открытия потока вывода
	
	mov	QWORD PTR -32[rbp], rax		# Вызов метода принт для строк
	lea	rdi, .LC11[rip]
	call	puts@PLT
	
	lea	rsi, -856[rbp]			# Запись адреса переменной в регистр
	lea	rdi, .LC1[rip]			# Запись строки в регистр
	mov	eax, 0
	call	__isoc99_scanf@PLT		# Вызов метода скан
	
	mov	eax, DWORD PTR -856[rbp]	# Запись команды в регистр
	cmp	eax, 1				# Сравнение команды с 1
	jne	.L17				# Если не равны прыжок на точку 17 
	mov	DWORD PTR -4[rbp], 0		# параметр цикла = 0
	jmp	.L18				# Прыжок на точку 18
.L19:
	mov	eax, DWORD PTR -4[rbp]		# Запись параметра в регистр
	mov	edx, DWORD PTR -848[rbp+rax*4]	# расчет адреса текущего элемента массива В
	mov	rdi, QWORD PTR -32[rbp]		# Запись указателя в регистр
	lea	rsi, .LC1[rip]			# Запись строки в регистр
	mov	eax, 0
	call	fprintf@PLT			# Вызов функции принт
	
	add	DWORD PTR -4[rbp], 1		# параметр цикла +=1
.L18:
	mov	eax, DWORD PTR -852[rbp]	# Запись сайз в регистр
	cmp	DWORD PTR -4[rbp], eax		# Сравнениие сайз с параметром
	jl	.L19				# Если параметр меньше, прыжок на 19
.L17:
	mov	eax, DWORD PTR -856[rbp]	# Получение значения команды
	cmp	eax, 2				# Сравнение с двойкой
	jne	.L20				# Если не равны то прыжок на точку 20
	mov	DWORD PTR -4[rbp], 0		# Парметр цикла  = 0
	jmp	.L21				# Прыжок на точку 21
.L22:
	mov	eax, DWORD PTR -4[rbp]		# Запись параметра в регистр
	
	mov	esi, DWORD PTR -848[rbp+rax*4]	# Расчет текущего элемента массива
	lea	rdi, .LC1[rip]			# Запись строки в регистр
	mov	eax, 0
	call	printf@PLT			# Вызов метода принт
	
	add	DWORD PTR -4[rbp], 1		# Параметр +=1
.L21:
	mov	eax, DWORD PTR -852[rbp]	# Запись сайз в регистр
	cmp	DWORD PTR -4[rbp], eax		# Сравнение сайз с параметром
	jl	.L22				# Прыжок на тчоку 22
.L20:
	mov	eax, 0
.L23:
	leave					# Выход из функции
	ret
	.size	main, .-main
